chapter2/monolithic_navierstokes_crp0.py

```diff
@@ -0,0 +1,219 @@
+    p.add_argument("--nx", type=int, default=64)
+    p.add_argument("--ny", type=int, default=32)
+    p.add_argument("--Lx", type=float, default=2.0)
+    p.add_argument("--Ly", type=float, default=1.0)
+    p.add_argument("--nu", type=float, default=1e-2)
+    p.add_argument("--uin", type=float, default=1.0)       
+    p.add_argument("--picard-it", type=int, default=20)
+    p.add_argument("--picard-tol", type=float, default=1e-8)
+    p.add_argument("--theta", type=float, default=0.5)
+    p.add_argument("--eps-p", type=float, default=1e-12)   # tiny pressure mass to remove nullspace
+    p.add_argument("--stokes-only", action="store_true")
+    p.add_argument("--no-output", action="store_true")
+    p.add_argument("--alpha", type=float, default=400.0, help="Nitsche penalty (100–1000 recommended)")
+    p.add_argument("--outfile", type=str, default="chapter2/open_cavity_crp0.xdmf")
+    return p.parse_args()
+
+
+args = parse_args()
+comm = MPI.COMM_WORLD
+rank = comm.rank
+
+# --- mesh ---
+domain = mesh.create_rectangle(
+    comm,
+    [np.array([0.0, 0.0]), np.array([args.Lx, args.Ly])],
+    [args.nx, args.ny],
+    cell_type=mesh.CellType.triangle,
+)
+tdim = domain.topology.dim
+fdim = tdim - 1
+gdim = domain.geometry.dim
+cell = domain.basix_cell()
+
+# Ensure needed connectivities
+domain.topology.create_connectivity(fdim, tdim)
+domain.topology.create_connectivity(tdim, tdim)
+
+# --- boundary facet tags (needed for ds with Nitsche) ---
+left_facets   = mesh.locate_entities_boundary(domain, fdim, lambda x: np.isclose(x[0], 0.0))
+right_facets  = mesh.locate_entities_boundary(domain, fdim, lambda x: np.isclose(x[0], args.Lx))
+bottom_facets = mesh.locate_entities_boundary(domain, fdim, lambda x: np.isclose(x[1], 0.0))
+top_facets    = mesh.locate_entities_boundary(domain, fdim, lambda x: np.isclose(x[1], args.Ly))
+
+# Tag: 1=left, 2=right, 3=bottom, 4=top
+facet_indices = np.concatenate([left_facets, right_facets, bottom_facets, top_facets]).astype(np.int32)
+facet_tags    = np.concatenate([
+    np.full_like(left_facets,   1, dtype=np.int32),
+    np.full_like(right_facets,  2, dtype=np.int32),
+    np.full_like(bottom_facets, 3, dtype=np.int32),
+    np.full_like(top_facets,    4, dtype=np.int32),
+])
+
+# Create facet meshtags
+if facet_indices.size == 0:
+    ft = mesh.meshtags(domain, fdim, np.array([], dtype=np.int32), np.array([], dtype=np.int32))
+else:
+    order = np.argsort(facet_indices)
+    ft = mesh.meshtags(domain, fdim, facet_indices[order], facet_tags[order])
+
+dx = ufl.Measure("dx", domain=domain)
+ds = ufl.Measure("ds", domain=domain, subdomain_data=ft)
+
+# --- spaces: CR–P0 (mixed) ---
+V_el = element("CR", cell, 1, shape=(gdim,))
+Q_el = element("DG", cell, 0)
+W = fem.functionspace(domain, mixed_element([V_el, Q_el]))
+(u, p) = ufl.TrialFunctions(W)
+(v, q) = ufl.TestFunctions(W)
+
+# --- parameters / RHS ---
+nu = fem.Constant(domain, PETSc.ScalarType(args.nu))
```

Comment:
```suggestion
nu = fem.Constant(domain, dolfinx.default_scalar_type(args.nu))
```
Use `dolfinx.default_scalar_type` rather than `PETSc.ScalarType` here an in all other places.